### PR TITLE
Feature/013 integrate new training platform implementation with gui

### PR DESCRIPTION
**Pull request by JanKapala**
Not entirely working, but there is significant progress, which make everyone else (especially Klooskie) able to work on his branch after merging these changes.

What have been done:
- Fixed restarting bug in gui:
     - made it non-blocking
     - fix EnvRestartedMsg handling by adding notify_on_end variables
- Fix blocking restart handling

Moreover:
- Now monitor is sorting messages by their sending(not receiving) timestamps even if they come in inapropriate order. It makes debuging based on logs much more comfortable
- Ad logger and loggin messages to:
     - EnvironmentServer,
     - GameManager,
     - TicTacToeComponent,
     - TicTacToeClient
- Handle "nothing to stop" case in stop_server script

TODO:
- Add restart handling to agents or agentClients to make episodes ultimately consistent

### CODE REVIEW

File: training_platform/server/service.py [hunk 1]
```diff
@@ -88,18 +100,28 @@ def receiveMessage(self, msg, sender):
...
             self.who_started_game = sender
+            self.log(f"New who_started_game={self.who_started_game}")
+            self.notify_on_end = msg.notify_on_end
+            self.log(f"New self.notify_on_end={self.notify_on_end}")
             self.environment.reset()
             for player in self.players_clients.keys():
                 self.before_first_move[player] = True
             for client in self.players_clients.values():
                 self.send(client, StateUpdateMsg(self.environment.current_board))
+            # if self.notify_on_end:
```
Comment (Klooskie) [inline training_platform/server/service.py]: smieciowy komentarz

File: game_app/games/tic_tac_toe/tic_tac_toe_component.py [hunk 2]
```diff
@@ -209,8 +223,13 @@ def step(self, position):
...
+        self.server.restart(blocking=False)
+        self.log("called: self.server.restart(blocking=False)")
 
     def back_to_menu(self):
         self.server.shutdown()
         self._app.switch_component(Components.MAIN_MENU)
+
+    def log(self, text):
+        if self.logger_addr is not None:
+            self.asys.tell(self.logger_addr, LogMsg(text, "TicTacToeComponent"))
```
Comment (Klooskie) [inline game_app/games/tic_tac_toe/tic_tac_toe_component.py]: brak entera

File: training_platform/clients/monitor_client.py [hunk 3]
```diff
@@ -7,10 +7,23 @@
...
 from training_platform.server.logger import Logger
+import heapq
+import numpy as np
+
+
+def clear():
+    if name == 'nt':
+        # for windows
+        system('cls')
+
```
Comment (Klooskie) [inline training_platform/clients/monitor_client.py]: po co ten enter

File: training_platform/clients/monitor_client.py [hunk 4]
```diff
@@ -19,6 +32,8 @@ def __init__(self):
         self.match_maker_addr = self.asys.createActor(MatchMaker, globalName="MatchMaker")
         self.game_manager_addr = self.asys.createActor(GameManager, globalName="GameManager")
         self.logger_addr = self.asys.createActor(Logger, globalName="Logger")
+        self.m_tuples = []
+
```
Comment (Klooskie) [inline training_platform/clients/monitor_client.py]: 2 entery zamiast 1

File: training_platform/clients/monitor_client.py [hunk 5]
```diff
@@ -28,4 +43,7 @@ def start_monitoring(self):
         while True:
             msg = self.asys.listen()
             if isinstance(msg, LogMsg):
-                print(msg)
+                heapq.heappush(self.m_tuples, (msg.time.timestamp(), np.random.rand(), msg))
+                clear()
+                for msg_tuple in heapq.nsmallest(len(self.m_tuples), self.m_tuples, lambda m_tuple: m_tuple[0]):
+                    print(msg_tuple[2])
```
Comment (Klooskie) [inline training_platform/clients/monitor_client.py]: enter na koncu

File: training_platform/server/environment_server.py [hunk 6]
```diff
@@ -52,29 +57,37 @@ def _connect(self):
...
+        response = self.asys.listen()
         if blocking:
             return self._start_blocking(response)
         else:
             return self._start_non_blocking(response)
 
     def _start_non_blocking(self, response):
         """Non-blocking call, returns immediately after EnvironmentServer has started environment"""
+        self.log(f"Received response: {response} in _start_non_blocking")
+        print(f"Received response: {response} in _start_non_blocking")
```
Comment (Klooskie) [inline training_platform/server/environment_server.py]: print w epoce logow

File: training_platform/server/service.py [hunk 7]
```diff
@@ -47,25 +49,33 @@ def receiveMessage(self, msg, sender):
...
+            self.log(f"Sent YourTurnMsg to current_client: {current_client}")
             self.send(self.who_started_game, EnvStartedMsg())
-
+            self.log(f"Sent EnvStartedMsg to self.who_started_game: {self.who_started_game}")
 
         elif isinstance(msg, TakeActionMsg):
+            self.log(f"Received TakeActionMsg")
             self.before_first_move[self.environment.current_player] = False
             self.environment.make_move(msg.action)  # It implicitly makes next player current player
             # State update for GUI clients TODO: send StateUpdateMsg only to GUI clients
```
Comment (Klooskie) [inline training_platform/server/service.py]: to juz jest zrobione z tego co widze wiec komentarz do wyrzucenia

File: training_platform/server/service.py (hunk 1 above)
Comment (JanKapala) [reply training_platform/server/service.py]: yep, usunę

File: training_platform/server/service.py (hunk 7 above)
Comment (JanKapala) [reply training_platform/server/service.py]: nie, na razie wysyła state update do wszystkich klientów, w tym RL, którzy po prostu to ignorują, chcę zrobić żeby StateUpdate był wysyłany tylko do klientów typu gui.

Comment (JanKapala) [reply training_platform/server/service.py]: Nie jest zrobione, póki co game manager wysyła StateUpdateMsg do wszystkich klientów zamiast tylko do GUI klientów

**Pull request merged**